src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -844,6 +826,41 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
             return EnumeratorResult.FailedNotReported;
         }
 
+        private void GetDisposalInfoForEnumerator(ref ForEachEnumeratorInfo.Builder builder, BoundExpression expr, bool isAsync, DiagnosticBag diagnostics)
+        {
+            // NOTE: if IDisposable is not available at all, no diagnostics will be reported - we will just assume that
+            // the enumerator is not disposable.  If it has IDisposable in its interface list, there will be a diagnostic there.
+            // If IDisposable is available but its Dispose method is not, then diagnostics will be reported only if the enumerator
+            // is potentially disposable.
+
+            TypeSymbol enumeratorType = builder.GetEnumeratorMethod.ReturnType.TypeSymbol;
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+
+            if (!enumeratorType.IsSealed ||
+                this.Conversions.ClassifyImplicitConversionFromType(enumeratorType,
+                    isAsync ? this.Compilation.GetWellKnownType(WellKnownType.System_IAsyncDisposable) : this.Compilation.GetSpecialType(SpecialType.System_IDisposable),
+                    ref useSiteDiagnostics).IsImplicit)
+            {
+                builder.NeedsDisposeMethod = true;
+            }
+            else if (enumeratorType.IsValueType && enumeratorType.IsRefLikeType)
+            {
+                // if it wasn't directly convertable to IDisposable, see if it is pattern disposable
+                // again, we throw away any binding diagnostics, and assume it's not disposable if we encounter errors
```

Comment:
If we find a `Dispose` method, but it is obsolete, that should be reported.

Reply:
Today we don't issue obsolete warnings for dispose methods, so pattern-based doesn't either ([SharpLab](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMACAwgWAFADe+GJGKAjAGxkAsGAsgIYwB2AFAJTGlF6n8Y4AeygRGAYwAWGNmEZQMADwysMLCAHcMAURYBXALbRGAIwA2Ecpy58BJXnbsUAdBQCcbRRwDc3OwF8/DEC8EPx0HX0jKFMLcnwHUhQAZgwAERgAZwAHIUzYiF1DY2ARDABxCGAi6MZSqE4MAjIAdjVNdKzc/PNCqJKRTm9g/DC8CIycvIKagYUQMnJnAElJ7oKEoNZgJV9bEhSVFh2sPShRY6aMAHMqq5Q2xWH/Ef2yVJMhITMGIUgAOQgimAjWaDwwAGoIcoADwYGjPIJBADaLgA8iZMt8qhAALpBQ4oOhrPIQUGLVzkDwAIjSQnUGA0MGA0hJPQssxi9WpPleISAA===))

I can add a test showing that neither version issues an obsolete warning?

Reply:
Yes please. Please link to an issue (we may have one already, I'm not sure)

Reply:
This is the issue [https://github.com/dotnet/roslyn/issues/31710](https://github.com/dotnet/roslyn/issues/31710)

---
In reply to: [250024707](https://github.com/dotnet/roslyn/pull/32640#discussion_r250024707) [](ancestors = 250024707)